spec/inputs/http_poller_spec.rb

```diff
@@ -2,87 +2,151 @@
-      # no stub, and an incorrect domain, this should result in no response.
+  let(:default_opts) {
+    {
+      "interval" => default_interval,
+      "urls" => default_urls,
+      "codec" => "json"
+    }
+  }
+  let(:mod) { LogStash::Inputs::HTTP_Poller }
 
-      logstash_queue = Queue.new
-      subject.run_once logstash_queue
+  subject { mod.new(default_opts) }
+
+  describe "#run" do
+    it "should run at the specified interval" do
+      expect(Stud).to receive(:interval).with(default_interval).once
+      subject.send(:run, double("queue"))
+    end
+  end
 
-      expect(logstash_queue.size).to eq(0)
+  describe "#run_once" do
+    it "should issue an async request for each url" do
+      default_urls.each { |name, url|
+        expect(subject).to receive(:request_async).with(queue, name, url).once
+      }
 
-      subject.teardown
+      subject.send(:run_once, queue)
     end
+  end
 
-    it "should create an event with an error" do
-      subject = LogStash::Inputs::HTTP_Poller.new(settings.merge({"urls" => {"url1" => "http://mytesturl.mydomain"}}))
-      subject.client.stub("http://mytesturl.mydomain", body:"not found",  code: 404)
+  shared_examples("matching metadata") {
+    let(:metadata) { event["@metadata"]["http_poller"] }
 
-      logstash_queue = Queue.new
-      subject.run_once logstash_queue
-      event = logstash_queue.pop
+    it "should have the correct name" do
+      expect(metadata["name"]).to eql(name)
+    end
+
+    it "should have the correct url" do
+      expect(metadata["url"]).to eql(url)
+    end
 
-      expect(event["responseCode"]).to eq(404)
-      expect(event["success"]).to eq(false)
-      subject.teardown
+    it "should have the correct code" do
+      expect(metadata["code"]).to eql(code)
     end
+  }
+
+  shared_examples "unprocessable_requests" do
+    let(:poller) {
+      p = LogStash::Inputs::HTTP_Poller.new(settings)
+      allow(p).to receive(:handle_failure).and_call_original
+      allow(p).to receive(:handle_success)
+      p
+    }
+    subject(:event) {
+      poller.send(:run_once, queue)
+      queue.pop(true)
+    }
+
+    before do
+      event # materialize the subject
+    end
+
+    it "should enqueue a message" do
+      expect(event).to be_a(LogStash::Event)
+    end
+
+    it "should enqueue a message with '_http_request_failure' set" do
+      expect(event["_http_request_failure"]).to be_a(Hash)
+    end
+
+    it "should tag the event with '_http_request_failure'" do
+      expect(event["tags"]).to include('_http_request_failure')
+    end
+
+    it "should invoke handle failure exactly once" do
+      expect(poller).to have_received(:handle_failure).once
+    end
+
+    it "should not invoke handle success at all" do
+      expect(poller).not_to have_received(:handle_success)
+    end
+
+    include_examples("matching metadata")
   end
 
-  context "with json codec, when remote server responds" do
+  describe "with a non responsive server" do
```

Comment:
my preference for prepositional phrases in rspec is to use the 'context' word instead of 'describe'

`context "with a non responsive server" do`

etc

Execution-wise, they are equivalent (describe and context are aliases of the same method in RSpec), but I feel the `context` word has better meaning given positional information (with ...) is context